### ISSUE TITLE
OneHasMany: find books with "Tag 1" or without tags

### ISSUE THREAD

**Issue by martenb** [nextras/orm#541]
**To Reproduce**
[PR with failing tests](https://github.com/nextras/orm/pull/540)

**Describe the bug**
*testHasValueOrEmptyWithNull*
- is ok in MySQL on 4.0.2, failing on master
- is failing in array on 4.0.2 and master

*testHasValueOrEmptyWithFunctions*
- is failing in MySQL on 4.0.2 and master
- is ok in array on 4.0.2 and master

**Comment by hrach** [comment]
Thanks for detailed report. Will take a look.

**Comment by hrach** [comment]
I have reviewed the test cases:

```php
$books = $this->orm->books->findBy([
	ICollection::OR,
	['tags->id' => [1]],
	['tags->id' => null],
]);

Assert::same(2, $books->countStored());
Assert::same(2, $books->count());
```

> - is ok in MySQL on 4.0.2, failing on master
> - is failing in array on 4.0.2 and master

The semantic meaning of this is:
"Select book that has tag with ID=1 or has tag with ID = null".
Since the second is never true, just the first condition is met -> the expected result is `1` on our testcase data-set.

I'd say this was buggy in Orm 4.0. With the proper expectation
```php
Assert::same(1, $books->countStored());
Assert::same(1, $books->count());
```
this is ok in 5.0-dev for **array** and will be fixed for DBAL's layers. This is currently an issue as we need to provide mechanism for auto aliasing repeated joins. It will be fixed by #526.

The alternative wanted behavior: "Select books with TAG#1 or with no tags" has to be written differently - as shown in the second failing testcase. Its DBAL mapper fix is coming with #526.

**Comment by hrach** [comment]
Hi, after weeks of exploring the implementation details & aspects I've finally manager to finish this great refactoring. 

The Nextras Orm 5 expected behavior is this:

```php
public function testSameTableJoinWithImplicitAggregation(): void
{
	$books = $this->orm->books->findBy([
		ICollection::OR,
		['tags->id' => [1]],
		['tags->id' => null], // no match
	]);

	Assert::same(1, $books->countStored());
	Assert::same(1, $books->count());
}
```

`Select book that has tag with ID=1 or has tag with ID = null`
Since the second condition is never true, just the first condition is met -> the expected result is 1 on our testcase data-set. This case is reusing the same "joins" - or rather aggregationKey as we call internally.

------------------

```php
public function testHasValueOrEmptyWithFunctions(): void
{
	$books = $this->orm->books->findBy([
		ICollection::OR,
		['tags->id' => [1]],
		[CompareEqualsFunction::class, [CountAggregateFunction::class, 'tags->id'], 0],
	]);

	Assert::same(2, $books->countStored());
	Assert::same(2, $books->count());
}
```
`Select book that has tag with ID=1 or that has count of tags equal to zero.`
The second condition is an aggregation over has many relationship and as such it obviously requires a different table join (an aggregation key) it the Orm implicitly does so! So this works now correctly for both array & dbal.

------------------

I'd discovered an issue in array implementation that evaluated all has-many relationships (joins) *independently* - in contrasts that's default behavior was "dependent" comparison (i.e. sharing the same join data).

```php
/*
 * Select author that has a book that:
 * - has title Book 1
 * - and is not translated.
 */
$authors = $this->orm->authors->findBy([
	ICollection::AND,
	'books->title' => 'Book 1',
	'books->translator->id' => null,
]);
$authors->fetchAll();
Assert::same(0, $authors->count());
Assert::same(0, $authors->countStored());
```

This previously worked differently in array & dbal. Now it properly filters out all authors in array. (The former behavior was not filtering author because he had an a book with title Book 1 and has a book that was not translated.)

------------------

Hopefully I didn't broke many other things.
Some things (grouping cols) are uselessly repeated. But this shouldn't matter much.
